### ISSUE TITLE
session.NewBatch(...).WithTimestamp(...) doesn't work

### ISSUE THREAD

**Issue by d3ce1t** [gocql/gocql#823]
It seems like custom timestamp set in a batch by using session.NewBatch(...).WithTimestamp(...) is lost or ignored somewhere.

I have created a test to check this behavior in -> https://play.golang.org/p/4U7bx564d5. Download it, change some constants to configure where is your Cassandra server and execute it.

This test connects to Cassandra, creates a new table in the keyspace you indicate and executes some queries that insert data in a configured table. More specifically, these tests are performed:

1) Insert two rows using session.Query(...).WithTimestamp()

2) Insert two rows using session.Query(...) and setting USING TIMESTAMP as part of the query.

3) Insert two rows in a batch using session.NewBatch(...).WithTimestamp(...). This is performed for LoggedBatch and UnloggedBatch.

4) Insert two rows in a batch using session.NewBatch(...) and setting USING TIMESTAMP as part of the query. This is performed for LoggedBatch and UnloggedBatch.

I don't know enough about gocql to figure it out where the problem  could be in the gocql code base.

**Comment by Zariel** [comment]
The protocol does not allow you to set a timestamp per query in a batch, only for the whole batch.

**Comment by d3ce1t** [comment]
I assume your comment refers to the tests Batch_Logged_UsingTimestamp and Batch_Unlogged_UsingTimestamp where I set USING TIMESTAMP as part of the two queries inside the batch. As you can see if run the test, timestamp set per query in the batch with USING TIMESTAMP works here.

As stated in https://docs.datastax.com/en/cql/3.3/cql/cql_reference/batch_r.html:

> BATCH allows you to set a client-supplied timestamp, an integer, in the USING TIMESTAMP clause. If you add USING TIMESTAMP but do not supply an integer value, Cassandra applies the time of the insertion (in microseconds).

> **The timestamp applies to all statements in the batch. You can apply a timestamp by adding a USING clause to any of the DML statements in the batch.**

Anyway, this test purpose is to stand out that both Batch_Logged_WithTimestamp and Batch_Unlogged_WithTimestamp don't work. In these tests session.NewBatch(...).WithTimestamp(...) is used to set the timestamp for the whole batch. However, when writetime is read back it doesn't match to the timestamp that was set by WithTimestamp(...). I presume this behaviour isn't the expected.

**Comment by Zariel** [comment]
Yes, but this is not supported in the protocol which is what WithTimestamp sets on the query, for batch see [0] and for query see [1]

[0] https://github.com/apache/cassandra/blob/22a3db0653ea5a45b396c7d8c07138d4257be5d6/doc/native_protocol_v4.spec#L379
[1] https://github.com/apache/cassandra/blob/22a3db0653ea5a45b396c7d8c07138d4257be5d6/doc/native_protocol_v4.spec#L307

**Comment by d3ce1t** [comment]
I don't think I understand your point or there is something I am missing most likely.

Do you mean that session.NewBatch(...).WithTimestamp(...) works by setting the timestamp of the individual queries at protocol level?. If that's the case, why not just setting the timestamp for the whole batch?

I have checked the specs you provided about CQL binary protocol. AFAIK batch message has the following body:

`<type><n><query_1>...<query_n><consistency><flags>[<serial_consistency>][<timestamp>]`

So timestamp for the whole message is there. Does session.NewBatch(...).WithTimestamp(...) make use of it? If so, why it doesn't pass the test I wrote? 

I think the session.NewBatch(...).WithTimestamp(...) is equivalent to the following CQL:

```
BEGIN BATCH
  USING TIMESTAMP timestamp
  dml_statement;
  dml_statement;
  ...
APPLY BATCH;
```

If not, what is the rationale behind that?

Regardless of the aforementioned, you have said several times that setting a timestamp to individual statements in a batch isn't supported at the protocol. I don't see where it's not supported. Let me explain myself.

A statement inside a batch has the following body:

`<kind><string_or_id><n>[<name_1>]<value_1>...[<name_n>]<value_n>`

where

```
<string_or_id> depends on the value of <kind>. If <kind> == 0, it should be
         a [long string] query string (as in QUERY, the query string might contain
         bind markers). Otherwise (that is, if <kind> == 1), it should be a
         [short bytes] representing a prepared query ID.
```

And a query string is like:

`<query><query_parameters>`

where 

`<query> is a [long string] representing the query`

and ***query_parameters*** are:

`<consistency><flags>[<n>[name_1]<value_1>...[name_n]<value_n>][<result_page_size>][<paging_state>][<serial_consistency>][<timestamp>]`

Assuming that ***query_parameters*** (which include the timestamp for the query) cannot be set to the DML statement inside a batch, I suppose "USING TIMESTAMP timestamp" would still be part the ***long string***, so it gets interpreted by the server. Am I wrong?.

This would explain why in [the tests I wrote](https://play.golang.org/p/4U7bx564d5) I can set timestamp to individuals queries inside a batch. Moreover, [I have updated the tests and added a new one](https://play.golang.org/p/X4iHx-5Gf5) that set a different timestamps to columns a and b in a single batch.

***Turning to the main question, I still don't understand why session.NewBatch(...).WithTimestamp(...) doesn't set the timestamp for the whole batch correctly.*** I can execute a batch in CQL from cqlsh with a provided timestamp and the timestamp is set as expected.

**Comment by Zariel** [comment]
Closing this as I think its fixed by #875
